### ISSUE TITLE
hsb colors and map-ing

### ISSUE THREAD

**Issue by alexandrdmitrievich** [processing/p5.js#1826]
https://alpha.editor.p5js.org/full/SkmmVoZ9x
Why does HSB coloring working differently?

**Comment by joecridge** [comment]
@alexandrdmitrievich could you post the code you’re using to draw the squares?

**Comment by alexandrdmitrievich** [comment]
function setup() { 
  createCanvas(400, 400);
} 

function draw() { 
 // noStroke();  // Don't draw a stroke around shapes
c = color('hsb('+map(mouseX,0,width,0,255)+',75%,100%)');
fill(c); 
rect(10, 10, 35, 35); 
  
c = color('hsb('+round(map(mouseX,0,width,0,255))+',75%,100%)');
fill(c); 
rect(10, 55, 35, 35); 

colorMode(HSB, 100);
c = color(map(mouseX,0,width,0,255),125,255);
fill(c);
rect(55, 10, 35, 35);  
}

**Comment by joecridge** [comment]
Okay, thanks.

Using an `'hsb(123, 45%, 67%)'` string to create colours only works [if the hue is an integer](https://github.com/processing/p5.js/blob/master/src/color/p5.Color.js#L378)\*, which is why the top left square is almost always white. Hue *always* goes from 0 to 360 (rather than 0 to 255) in these strings, which is why you don’t see the full spectrum in the bottom left square.

When you set `colorMode(HSB, 100)` before drawing the right square, you make it so that hue (*as a number*, going straight into `color()`) goes from 0 to 100, so you see even less of the spectrum. Additionally,  you are making it so that saturation and lightness also go from 0 to 100 (again, *as numeric arguments*), so the 125 and the 255 both get maxed out at 100 behind the scenes.

Note also that the colour mode persists once it has been set, so after the first draw loop the colour mode is always HSB (because you never change it to anything else), but this doesn’t affect the left two rectangles because *calling `color()` with a string always ignores `colorMode()`*.

---

Putting that all together, you would need something like this (to get the bottom left square to match the right one, and to make them both show the full spectrum):

```javascript
function setup() {
  createCanvas(400, 400);

  // We can set `colorMode()` up here because it’s the same every loop.
  // This is the same as calling `colorMode(HSB, 360, 100, 100)`.
  colorMode(HSB);
}

function draw() {
  // Hue in strings goes from 0 to 360.
  // This one will be white when string hue is decimal…
  c = color('hsb(' + map(mouseX, 0, width, 0, 360) + ', 75%, 100%)');
  fill(c);
  rect(10, 10, 35, 35);

  // …but this one should be drawn as expected.
  c = color('hsb(' + round(map(mouseX, 0, width, 0, 360)) + ', 75%, 100%)');
  fill(c);
  rect(10, 55, 35, 35);

  // Hue as a number goes from 0 to 360 in the default HSB mode.
  // Saturation and brightness go from 0 to 100.
  // Decimals are fine here.
  c = color(map(mouseX, 0, width, 0, 360), 75, 100);
  fill(c);
  rect(55, 10, 35, 35);
}
```

\* *@lmccart is this the desired behaviour? I believe CSS supports decimal hue.*

**Comment by lmccart** [comment]
@mlarghydracept do you want to take a look at this issue mentioned in the edit above?

**Comment by lmccart** [comment]
closed with #1891
